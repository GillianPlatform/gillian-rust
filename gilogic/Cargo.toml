[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }