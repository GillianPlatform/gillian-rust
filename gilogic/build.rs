fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
