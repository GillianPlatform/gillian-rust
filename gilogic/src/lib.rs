#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
