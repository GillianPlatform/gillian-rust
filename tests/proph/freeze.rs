//@ check-pass
extern crate creusillian;
extern crate gilogic;

use gilogic::{
    macros::*,
    prophecies::{Ownable, Prophecised, Prophecy},
};

struct A {
    x: u64,
    y: u64,
}

#[with_freeze_lemma(
    lemma_name = freeze_x,
    predicate_name = a_ref_mut_x,
    frozen_variables = [x]
)]
impl Ownable for A {
    type RepresentationTy = u64;

    #[predicate]
    fn own(self, model: Self::RepresentationTy) {
        assertion!(|x: u64, y| (self == A { x, y }) * (model == x) * (y == x + 1))
    }
}
