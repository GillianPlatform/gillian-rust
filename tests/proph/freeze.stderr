warning: unused imports: `Prophecised`, `Prophecy`
 --> ../tests/proph/freeze.rs:7:27
  |
7 |     prophecies::{Ownable, Prophecised, Prophecy},
  |                           ^^^^^^^^^^^  ^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

[TIMESTAMP WARN  lib_rtg::metadata] could not read metadata for crate `"std"`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
[TIMESTAMP WARN  lib_rtg::metadata] could not read metadata for crate `"creusillian"`: Os { code: 2, kind: NotFound, message: "No such file or directory" }
warning: fields `x` and `y` are never read
  --> ../tests/proph/freeze.rs:11:5
   |
10 | struct A {
   |        - fields in this struct
11 |     x: u64,
   |     ^
12 |     y: u64,
   |     ^
   |
   = note: `#[warn(dead_code)]` on by default

warning: variable `THIS` should have a snake case name
  --> ../tests/proph/freeze.rs:24:12
   |
24 |     fn own(self, model: Self::RepresentationTy) {
   |            ^^^^ help: convert the identifier to snake case: `this`
   |
   = note: `#[warn(non_snake_case)]` on by default

warning: 3 warnings emitted

[TIMESTAMP INFO  rust_to_gil] Compiled!
